
<font size="7">__Leaf Identification Model__</font>


## Project Description
<img src='Leaves.jpg' width=800 height=600/></a>
In this Machine Learning project, there are six sets of totally different images. All sets have 100 images each of Cherry Blossom, Gingko, Japanese Maple, Red Oak, Silver Birch and Sweet Gum leaves. All these images are in the Folder, /Images.

A model was built such that it is capable of recognizing an unknown leaf and distinguish it accurately.<br>

There is also a Test_Set, containing six Folders each containing images of differenr type of leaves. So, model accuracy can be evaluated with this test_set. Later, model will be used to predict if a single unknown image is of what specie of leaf.<br>


In this exercise our objective is to perform the following tasks:

1. Import the set of images
2. Building an appropriate CNN model
3. Training the CNN model
4. Making Prediction off the trained model
7. Checking Accuracy of Model

## Importing the Libraries

In [1]:
from google.colab import drive
drive.mount('/content/gdrive')

ModuleNotFoundError: No module named 'google.colab'

In [ ]:
import tensorflow as tf
from keras.preprocessing.image import ImageDataGenerator

## Data Preprocessing

### Preprocessing the Training Set

In [ ]:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)
training_set = train_datagen.flow_from_directory('/content/gdrive/MyDrive/Colab Notebooks/Leaves/Images/Train',
                                                 target_size = (64, 64),
                                                 batch_size = 32,
                                                 class_mode = 'categorical')

Found 420 images belonging to 6 classes.


### Preprocessing the Test Set

In [ ]:
test_datagen = ImageDataGenerator(rescale = 1./255)
test_set = test_datagen.flow_from_directory('/content/gdrive/MyDrive/Colab Notebooks/Leaves/Images/Test',
                                            target_size = (64, 64),
                                            batch_size = 32,
                                            class_mode = 'categorical')

Found 190 images belonging to 6 classes.


## Building the CNN Model

### Step 1 - Initialising the Model

In [ ]:
Model = tf.keras.models.Sequential()

### Step 2 - Adding First Convolution Layer 
The number of filters directly affects the accuracy level of the model, increasing it will increase the accuracy.

In [ ]:
Model.add(tf.keras.layers.Conv2D(filters=10, kernel_size=5, activation='relu', input_shape=[64, 64, 3]))

### Step 3 - Pooling the First Layer

In [ ]:
Model.add(tf.keras.layers.MaxPool2D(pool_size=4, strides=2))

### Step 4 - Adding a Second Convolutional Layer

In [ ]:
Model.add(tf.keras.layers.Conv2D(filters=10, kernel_size=5, activation='relu'))

### Step 5 - Pooling the Second Layer

In [ ]:
Model.add(tf.keras.layers.MaxPool2D(pool_size=4, strides=2))

### Step 6 - Flattening

In [ ]:
Model.add(tf.keras.layers.Flatten())

### Step 7 - Full Connection

In [ ]:
Model.add(tf.keras.layers.Dense(units=128, activation='relu'))

### Step 8 - Output Layer
For classes greater than two, a softmax activation function is used

In [ ]:
Model.add(tf.keras.layers.Dense(units=6, activation='softmax')) # units = number of classes

### Step 9 - Compiling the CNN
In case of N-classes, categorical crossentropy method is used, with nadam optimizer

In [ ]:
Model.compile(loss='categorical_crossentropy', optimizer='nadam',metrics=['accuracy'])

## Training the CNN and Evaluation

In [ ]:
Model.fit(x = training_set, validation_data = test_set, epochs = 25) #check with 20, 25 and 30

Epoch 1/25
14/14 [==============================] - 351s 25s/step - loss: 1.7234 - accuracy: 0.2690 - val_loss: 1.6037 - val_accuracy: 0.4842
Epoch 2/25
14/14 [==============================] - 196s 14s/step - loss: 1.3679 - accuracy: 0.4714 - val_loss: 1.2855 - val_accuracy: 0.5158
Epoch 3/25
14/14 [==============================] - 193s 14s/step - loss: 0.9706 - accuracy: 0.6524 - val_loss: 2.7800 - val_accuracy: 0.1737
Epoch 4/25
14/14 [==============================] - 193s 14s/step - loss: 0.9599 - accuracy: 0.6833 - val_loss: 0.8053 - val_accuracy: 0.7684
Epoch 5/25
14/14 [==============================] - 196s 15s/step - loss: 0.5919 - accuracy: 0.7929 - val_loss: 1.1313 - val_accuracy: 0.5737
Epoch 6/25
14/14 [==============================] - 194s 14s/step - loss: 0.4893 - accuracy: 0.8381 - val_loss: 0.6298 - val_accuracy: 0.8158
Epoch 7/25
14/14 [==============================] - 192s 14s/step - loss: 0.4335 - accuracy: 0.8571 - val_loss: 0.5415 - val_accuracy: 0.8579
Epoch 

## Making a Prediction

In [2]:
import numpy as np
import os
from keras.preprocessing import image
import keras.utils
result =[]
path = '/content/gdrive/MyDrive/Colab Notebooks/Leaves/Unkown/Leaf'
for img in os.listdir(path):
    test_image = keras.utils.load_img(path+"/"+img, target_size = (64, 64))
    test_image = keras.utils.img_to_array(test_image)
    test_image = np.expand_dims(test_image, axis = 0)
    result.append(Model.predict(test_image))
    predict = Model.predict(test_image)
    #print (result[i][0][0])
    if predict[0][0] == 1:
        prediction = 'Cherry Blossom'
    elif predict[0][1] == 1:
        prediction = 'Gingko'
    elif predict[0][2] == 1:
        prediction = 'Japanese Maple'
    elif predict[0][3] == 1:
        prediction = 'Red Oak'
    elif predict[0][4] == 1:
        prediction = 'Silver Birch'
    elif predict[0][5] == 1:
        prediction = "Sweet Gum"
    print(img + " = "+ prediction)

ModuleNotFoundError: No module named 'tensorflow.python'

In [ ]:
test_image = keras.utils.load_img('/content/gdrive/MyDrive/Colab Notebooks/Leaves/Unkown/Leaf/Leaf_12.jpg', target_size = (64, 64))
test_image = keras.utils.img_to_array(test_image)
test_image = np.expand_dims(test_image, axis = 0)
predicting = Model.predict(test_image)
training_set.class_indices
if predict[0][0] == 1:
    prediction = 'Cherry Blossom'
elif predict[0][1] == 1:
    prediction = 'Gingko'
elif predict[0][2] == 1:
    prediction = 'Japanese Maple'
elif predict[0][3] == 1:
    prediction = 'Red Oak'
elif predict[0][4] == 1:
    prediction = 'Silver Birch'
elif predict[0][5] == 1:
    prediction = "Sweet Gum"
print(prediction)

1/1 [==============================] - 0s 19ms/step
Sweet Gum


## Model Accuracy

In [ ]:
Model.evaluate(training_set)

14/14 [==============================] - 135s 10s/step - loss: 0.0515 - accuracy: 0.9833


[0.05151832848787308, 0.9833333492279053]

In [ ]:
Model.evaluate(test_set)

6/6 [==============================] - 66s 11s/step - loss: 0.2244 - accuracy: 0.9421


[0.22438454627990723, 0.942105233669281]

The CNN classification model was found to be 98.33 % accurate on training set and 94.21 % accurate on validation set. 

##### End